apps/application-system/api/src/app/modules/application/application.controller.ts

```diff
@@ -639,70 +643,90 @@ export class ApplicationController {
-    )
-    const intl = await this.intlService.useIntl(namespaces, locale)
+      for (let i = 0; i < externalDataDto.dataProviders.length; i++) {
+        const found = providersFromRole.find(
+          (x) => x.actionId === externalDataDto.dataProviders[i].actionId,
+        )
 
-    const templateApis: TemplateApi[] = []
+        if (found) {
+          templateApis.push(found)
+        } else {
+          throw new BadRequestException(
+            `Current user is not permitted to update external data in this state with actionId: ${externalDataDto.dataProviders[i].actionId}`,
+          )
+        }
+      }
 
-    for (let i = 0; i < externalDataDto.dataProviders.length; i++) {
-      const found = providersFromRole.find(
-        (x) => x.actionId === externalDataDto.dataProviders[i].actionId,
+      await this.validationService.validateIncomingExternalDataProviders(
+        existingApplication as BaseApplication,
+        templateApis,
+        user.nationalId,
       )
 
-      if (found) {
-        templateApis.push(found)
-      } else {
-        throw new BadRequestException(
-          `Current user is not permitted to update external data in this state with actionId: ${externalDataDto.dataProviders[i].actionId}`,
+      const updatedApplication = await this.templateApiActionRunner.run(
+        existingApplication as BaseApplication,
+        templateApis,
+        user,
+        locale,
+        intl.formatMessage,
+      )
+
+      if (!updatedApplication) {
+        throw new NotFoundException(
+          `An application with the id ${id} does not exist`,
         )
       }
-    }
 
-    await this.validationService.validateIncomingExternalDataProviders(
-      existingApplication as BaseApplication,
-      templateApis,
-      user.nationalId,
-    )
+      this.auditService.audit({
+        auth: user,
+        action: 'updateExternalData',
+        resources: existingApplication.id,
+        meta: { providers: externalDataDto },
+      })
 
-    const updatedApplication = await this.templateApiActionRunner.run(
-      existingApplication as BaseApplication,
-      templateApis,
-      user,
-      locale,
-      intl.formatMessage,
-    )
+      return updatedApplication
+    })
+  }
 
-    if (!updatedApplication) {
-      throw new NotFoundException(
-        `An application with the id ${id} does not exist`,
-      )
+  private logSubmissionResult(
+    applicationId: string,
+    templateId: string,
+    hasError = false,
+  ) {
+    const context = {
+      applicationId,
+      templateId,
     }
 
-    this.auditService.audit({
-      auth: user,
-      action: 'updateExternalData',
-      resources: existingApplication.id,
-      meta: { providers: externalDataDto },
+    withLoggingContext(context, () => {
+      if (hasError) {
+        this.logger.error(`Application submission ended with an error`)
+      } else {
+        this.logger.info(`Application submission ended successfully`)
+      }
```

Comment:
~Fyi, for a simple log like this, it's better to just add the context directly:

```ts
      if (hasError) {
        this.logger.error(`Application submission ended with an error`, context)
      } else {
        this.logger.info(`Application submission ended successfully`, context)
      }
```

`withLoggingContext` is ideal for a large tree of nested function calls, where you don't know what logs are being performed.